tests/core/test_datamodules.py

```diff
@@ -533,6 +534,56 @@ def test_simple_hyperparameters_saving():
     assert data.hparams == AttributeDict({"arg0": 10, "arg1": "foo", "kwarg0": "bar"})
 
 
+def test_define_as_dataclass():
+    # makes sure that no functionality is broken and the user can still manually make
+    # super().__init__ call with parameters
+    # also tests all the dataclass features that can be enabled without breaking anything
+    @dataclass(init=True, repr=True, eq=True, order=True, unsafe_hash=True, frozen=False)
+    class BoringDataModule1(LightningDataModule):
+        batch_size: int = 32
+        dims: int = 2
+
+        def __post_init__(self):
+            super().__init__(dims=self.dims)
+
+    # asserts for the different dunder methods added by dataclass, when __init__ is implemented, i.e.
+    # __repr__, __eq__, __lt__, __le__, etc.
+    assert BoringDataModule1(batch_size=32)
+    assert hasattr(BoringDataModule1, "__repr__")
+    assert BoringDataModule1(batch_size=32) == BoringDataModule1(batch_size=32)
+
+    # asserts inherent calling of super().__init__ in case user doesn't make the call
+    @dataclass
+    class BoringDataModule2(LightningDataModule):
+        batch_size: int = 2
+
+    # asserts for the different dunder methods added by dataclass, when super class is inherently initialized, i.e.
+    # __init__, __repr__, __eq__, __lt__, __le__, etc.
+    assert BoringDataModule2(batch_size=32).prepare_data() is None
+    assert BoringDataModule2(batch_size=32)
+    assert hasattr(BoringDataModule2, "__repr__")
+    assert BoringDataModule2(batch_size=32) == BoringDataModule2(batch_size=32)
+
+    # checks that empty super().__init__ call isn't made in case LightningDataModule isn't the direct parent
+    # of the defined class.
+    @dataclass
+    class BoringDataModule3(SklearnDataModule):
+        num_features: int = 32
+        length: int = 800
+        num_classes: int = 3
+        batch_size: int = 2
+
+        def __post_init__(self):
+            super().__init__(
```

Comment:
For this class, `mro[mro.index(cls) + 1] == LightningDataModule` will be False correct?

Then does that mean with multiple inheritance levels, the dataclass is forced to call `__init__` in its `__post_init__`?

Reply:
Yes. The class directly inheriting from `LightningDataModule` in that case will have to make `super().__init__` call. 
 
To make it so that all the following cases pass:
![Screenshot 1](https://user-images.githubusercontent.com/40473118/131465451-b2103927-1478-4c4d-8b03-bfaebc021ccd.png)

there could be one solution. We can replace:
```
mro = cls.mro()
if mro[mro.index(cls) + 1] == LightningDataModule:
    super(cls, obj).__init__()
```
with just:
```
LightningDataModule.__init__(obj)
```

This would ensure that LightningDataModule is initialized, even if any further derived class is initialized or not. And can be easily overridden by a second `__init__` call.

It shouldn't break anything, would it?

Reply:
I don't think so, try pushing that! :)

Reply:
Have pushed the changes. Can you please allow the workflows?

Reply:
Works well and good. All types of multilevel inheritance cases pass, along with all the tests for direct inheritance with `dataclass`. :D